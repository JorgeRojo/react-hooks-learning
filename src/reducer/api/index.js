import _get from 'lodash/get';

import {
  API_ERROR,
  API_REQUEST,
  API_RESET,
  API_SUCCESS,
} from '../../actions/actionTypes';

const defaultState = {
  response: {},
  isLoading: false,
};

const apiReducer = (state = defaultState, action) => {
  switch (action.type) {
    case API_RESET:
      return defaultState;
    case API_REQUEST:
      return { ...defaultState, isLoading: true };
    case API_ERROR:
      return { ...defaultState, isLoading: false };
    case API_SUCCESS:
      return {
        response: action.payload.response,
        isLoading: false,
      };
    default:
      return state;
  }
};

export default apiReducer;

// Selectors

export const selectApiData = state => _get(state, 'response.data', []);

export const selectApiIsLoading = state => state.isLoading;
