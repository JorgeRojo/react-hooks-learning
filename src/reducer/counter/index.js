import {
  COUNTER_DECREASE,
  COUNTER_INCREASE,
  COUNTER_RESET,
  COUNTER_SET
} from "../../actions/actionTypes";

const defaultState = 0;

const counterReducer = (state = defaultState, action) => {
  switch (action.type) {
    case COUNTER_DECREASE:
      return (state = state - 1);
    case COUNTER_INCREASE:
      return (state = state + 1);
    case COUNTER_RESET:
      return defaultState;
    case COUNTER_SET:
      return action.payload.number;
    default:
      return state;
  }
};

export default counterReducer;

export const selectCounter = state => state;
