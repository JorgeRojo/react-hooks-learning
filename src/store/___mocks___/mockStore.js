import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import mockState from "./mockState";
import withReduxDispatchGenerator from "../withReduxDispatchGenerator";

const middleware = [thunk];
const mockStore = configureMockStore(middleware);

const store = mockStore(mockState);

export default store;

export const withReduxDispatch = withReduxDispatchGenerator(store);
