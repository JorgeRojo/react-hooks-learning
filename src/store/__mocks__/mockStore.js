import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import mockState from './mockState';

const middleware = [thunk];
const mockStore = configureMockStore(middleware);

const store = mockStore(mockState);

export default store;
