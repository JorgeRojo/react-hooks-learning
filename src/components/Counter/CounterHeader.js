import React from 'react';
import { connect } from 'react-redux';
import { selectCounter } from '../../reducer';

const CounterHeader = ({ counter }) => {
  return <h1>Counter: {counter}</h1>;
};

const mapStateToProps = state => ({
  counter: selectCounter(state),
});

export default connect(mapStateToProps)(CounterHeader);
