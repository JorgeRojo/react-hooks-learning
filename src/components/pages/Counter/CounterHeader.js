import React from 'react';
import { connect } from 'react-redux';
import { selectCounter } from '../../../reducer';

const CounterHeader = ({ counter }) => {
  return <h1 className="display-4 mb-4">Counter: {counter}</h1>;
};

const mapStateToProps = state => ({
  counter: selectCounter(state),
});

export default connect(mapStateToProps)(CounterHeader);
