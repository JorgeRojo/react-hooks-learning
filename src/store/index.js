import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from 'redux-thunk';
import reducer from '../reducer';

const initialState = {};
const middleware = [thunk];
const composeEnhancers = composeWithDevTools({});

const store = createStore(
  reducer,
  initialState,
  composeEnhancers(applyMiddleware(...middleware))
);

export default store;
