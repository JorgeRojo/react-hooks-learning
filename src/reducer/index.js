import { combineReducers } from 'redux';
import api from './api';
import counter from './counter';

import * as fromCounter from './counter';
import * as fromApi from './api';

const reducer = {
  api,
  counter,
};

export default combineReducers(reducer);

// Selectors

export const selectCounter = state => fromCounter.selectCounter(state.counter);

export const selectApiData = state => fromApi.selectApiData(state.api);

export const selectApiIsLoading = state =>
  fromApi.selectApiIsLoading(state.api);
