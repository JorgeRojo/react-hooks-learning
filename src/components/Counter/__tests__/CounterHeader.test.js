import React from "react";
import ReactDOM from "react-dom";
import { render, cleanup } from "@testing-library/react";
import { Provider } from "react-redux";

import store from "../../../store";

import { selectCounter } from "../../../reducer/counter";

import CounterHeader from "../CounterHeader";

jest.mock("../../../store", () =>
  require("../../../store/___mocks___/mockStore")
);
jest.mock("../../../reducer/counter");

describe("CounterHeader Component", () => {
  const component = (
    <Provider store={store}>
      <CounterHeader />
    </Provider>
  );

  beforeEach(() => {
    selectCounter.mockReturnValue(999);
  });

  afterEach(() => {
    jest.clearAllMocks();
    cleanup();
  });

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(component, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders the component", () => {
    const { container } = render(component);
    expect(container).toMatchSnapshot();
  });
});
