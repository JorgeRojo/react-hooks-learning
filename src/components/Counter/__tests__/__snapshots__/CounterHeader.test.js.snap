// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterHeader Component renders the component 1`] = `
<div>
  <h1>
    Counter: 
    999
  </h1>
</div>
`;
